ecommerce/views_test.py

```diff
@@ -1384,6 +1384,54 @@ def test_products_viewset_detail(user_drf_client, coupon_product_ids):
     )
 
 
+def test_products_viewset_expired_programs(user_drf_client):
+    """ Test that the ProductViewSet returns contains only valid programs products and exclude the expired programs correctly"""
+    now = now_in_utc()
+    programs = ProgramFactory.create_batch(3)
+    runs = CourseRunFactory.create_batch(2, course__program=factory.Iterator(programs))
+    index = 0
+    while index < len(programs):
+        ProgramRunFactory.create(
+            program=programs[index],
+            start_date=(now + timedelta(days=1))
+            if index < (len(programs) - 1)
+            else (now - timedelta(days=3)),
+            end_date=(now + timedelta(days=2))
+            if index < (len(programs) - 1)
+            else (now - timedelta(days=1)),
+        )
+        index += 1
+
+    ProductVersionFactory.create_batch(
+        4, product__content_object=factory.Iterator(runs + programs)
+    )
```

Comment:
I don't think this test is testing something it really should with the past/expired programs. Instead the test is currently working on a missing version instead of a past program. This is one main issue in the test. We should be creating 5 ProductVersion objects. 

To see why this test is currently not working as it should, you'll need to make a tweak in the view to fail the test. Just replace 
```        expired_programs = (
            Program.objects.filter(
                Q(programruns__isnull=True)
                | Q(id__in=unused_program_id)
                | Q(courses__in=expired_courses)
            )
            .values_list("id", flat=True)
            .distinct()
        )
``` 
with 
```
expired_programs = Program.objects.filter(id=-1)
```
So that the view doesn't care for the expired programs. When you run the test it will still pass whereas it should fail now since expired programs are returned from view. This is happening because the test is working on missing product version instead.

```suggestion
    ProductVersionFactory.create_batch(
        5, product__content_object=factory.Iterator(runs + programs)
    )
```

Reply:
@arslanashraf7 , I forgot to update the count in `ProductVersionFactory` here in test from 4 to 5, (2 course runs and 3 programs) but it does not mean, the test logic is incorrect. 

In earlier written test, Someone forgot to handle to logic for expired programs due to that the logic in view for `expired_programs` always return an empty list and test passed but here in test, the purpose is to exclude the program that is in past and return the non-expired programs that the view logic with test is returning correctly. 

Anyways, Thanks for pointing out the increase in count while creating `ProductVersion` here.

Reply:
This assert [statement](https://github.com/mitodl/mitxpro/pull/2379/files#diff-8eed442c6f918427c359555eb169aeeeabb4f6c4b903c7bff2598075af82627eR1422) is making it sure